flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultExecutionSlotAllocatorTest.java

```diff
@@ -0,0 +1,355 @@
+		executionSlotAllocator.cancel(inValidExecutionVertexId);
+		assertThat(cancelledSlotRequestIds, hasSize(0));
+
+		assertThat(receivedSlotRequestIds, hasSize(executionVerticesNum / 2));
+
+		completeProducerSlotRequest(backupSlotFutures);
+
+		for (ExecutionVertexID executionVertexId : executionVertexIds) {
+			executionSlotAllocator.cancel(executionVertexId);
+		}
+
+		// only the consumers slot request will be cancelled
+		List<SlotRequestId> expectCancelledSlotRequestIds =
+				receivedSlotRequestIds.subList(executionVerticesNum / 2, executionVerticesNum);
+		assertThat(cancelledSlotRequestIds, containsInAnyOrder(expectCancelledSlotRequestIds.toArray()));
+	}
+
+	/**
+	 * Tests that all unfulfilled slot requests will be cancelled when stopped.
+	 */
+	@Test
+	public void testStop() {
+		executionSlotAllocator.allocateSlotsFor(schedulingRequirements);
+
+		// only the producers will ask for slots as the preferred location futures of consumers are not fulfilled.
+		assertThat(executionSlotAllocator.getPendingSlotAssignments().keySet(), hasSize(executionVerticesNum));
+		assertThat(receivedSlotRequestIds, hasSize(executionVerticesNum / 2));
+
+		executionSlotAllocator.stop().getNow(null);
+
+		assertThat(cancelledSlotRequestIds, hasSize(executionVerticesNum / 2));
+		assertThat(cancelledSlotRequestIds, containsInAnyOrder(receivedSlotRequestIds.toArray()));
+		assertThat(executionSlotAllocator.getPendingSlotAssignments().keySet(), hasSize(0));
+	}
+
+	/**
+	 * Tests that all prior allocation ids are computed by union all previous allocation ids in scheduling requirements.
+	 */
+	@Test
+	public void testComputeAllPriorAllocationIds() {
+		List<AllocationID> expectAllocationIds = Arrays.asList(new AllocationID(), new AllocationID());
+		List<ExecutionVertexSchedulingRequirements> testSchedulingRequirements = Arrays.asList(
+				new ExecutionVertexSchedulingRequirements(
+						executionVertexIds.get(0),
+						expectAllocationIds.get(0),
+						ResourceProfile.UNKNOWN,
+						null,
+						null,
+						null),
+				new ExecutionVertexSchedulingRequirements(
+						executionVertexIds.get(1),
+						expectAllocationIds.get(0),
+						ResourceProfile.UNKNOWN,
+						null,
+						null,
+						null),
+				new ExecutionVertexSchedulingRequirements(
+						executionVertexIds.get(2),
+						expectAllocationIds.get(1),
+						ResourceProfile.UNKNOWN,
+						null,
+						null,
+						null),
+				new ExecutionVertexSchedulingRequirements(
+						executionVertexIds.get(3),
+						null,
+						ResourceProfile.UNKNOWN,
+						null,
+						null,
+						null)
+		);
+
+		Set<AllocationID> allPriorAllocationIds = executionSlotAllocator.computeAllPriorAllocationIds(testSchedulingRequirements);
+		assertThat(allPriorAllocationIds, containsInAnyOrder(expectAllocationIds.toArray()));
+	}
+
+	/**
+	 * Tests the calculation of preferred locations based on inputs for an execution.
+	 */
+	@Test
+	public void testGetPreferredLocationsBasedOnInputs() throws Exception {
```

Comment:
This test is too big and violates the rule of having only a _single concept per test_ (see Clean Code by Robert C. Martin, Chapter 9, pp. 130, 131), i.e., this should be broken down to multiple tests. I would even prefer to move these into a separate class since we are testing a static method on `DefaultExecutionSlotAllocator` and not an instance of it.

I think there are (at least) two test cases needed:
- test that we do not wait for the futures if we have more than MAX_DISTINCT_LOCATIONS_TO_CONSIDER futures
- test that we prefer the TM locations of the predecessor with fewest distinct TM locations

Reply:
Good idea